File: aoc-py/solutions/day16.py
"""
Day 16: Reindeer Maze

https://adventofcode.com/2024/day/16
"""
__all__ = ('Day16',)

from typing import ClassVar
from heapq import heappop, heappush

from ..solution import Solution
from ..utils import find_start, neighbors_4

class Day16(Solution):
    NAME: ClassVar[str] = 'Reindeer Maze'

    def __init__(self) -> None:
        self.score = None
        self.seats = None

    def _traverse_maze(self, inp: str) -> tuple[int, int]:
        grid = inp.splitlines()
        start = find_start(grid, 'S')

        heap: list[tuple[int, int, int, int, int, tuple[tuple[int, int], ...]]] = []
        heappush(heap, (0, *start, 0, 1, ()))

        scores = {}
        all_seats = {start}

        min_score = 0

        while heap:
            score, row, col, dr, dc, seats = heappop(heap)

            if grid[row][col] == 'E':
                # completed an optimal path to the end
                # union the seats (tiles traversed) in the current path to all of the seats
                all_seats.update(seats)
                # Part 1 answer
                min_score = score
                continue

            for next_row, next_col in neighbors_4(row, col):
                # re-calculate the `dr`` and `dc` to get to the next tile
                next_dr = next_row - row
                next_dc = next_col - col

                # hit a wall or is attempting to go backwards
                if grid[next_row][next_col] == '#' or next_dr == -dr and next_dc == -dc:
                    continue

                # new score: `+1` if straight else `+1001` (`1000` for turning & `1` for moving again in the new direction)
                next_score = score + (
                    1 if next_dr == dr and next_dc == dc else 1001
                )

                # if the next tile has never been traversed with the same direction
                # OR if it has been: the current score is lower and more optimal
                #
                # -> re-traverse
                if scores.get(key := (next_row, next_col, next_dr, next_dc), next_score) >= next_score:
                    scores[key] = next_score
                    heappush(
                        heap,
                        (next_score, *key, seats + ((next_row, next_col),))
                    )

        return min_score, len(all_seats)

    def part_one(self, inp: str) -> int:
        if self.score is None:
            self.score, self.seats = self._traverse_maze(inp)
        return self.score

    def part_two(self, inp: str) -> int:
        if self.seats is None:
            self.score, self.seats = self._traverse_maze(inp)
        return self.seats

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 105496
        assert p2 == 524

File: src/bin/day16.rs
//! Day 16: Reindeer Maze
//!
//! <https://adventofcode.com/2024/day/16>

use std::{
    collections::{BinaryHeap, HashMap, HashSet},
    fmt::Display,
    cmp::Reverse,
};
use aoc_2024::{Solution, neighbors_4, find_start, get_grid};

pub struct Day16;

impl Day16 {
    fn traverse_maze<T: Display>(inp: T, is_part2: bool) -> usize {
        let grid = get_grid(inp);
        let start @ (start_row, start_col) = find_start(&grid, b'S').unwrap();

        let mut heap = BinaryHeap::new();
        heap.push(Reverse(
            (0, start_row, start_col, 0, 1, Vec::new())
        ));

        let mut scores = HashMap::new();
        let mut all_seats = HashSet::new();
        all_seats.insert(start);

        while let Some(Reverse((score, row, col, dr, dc, seats))) = heap.pop() {
            if grid[row][col] == b'E' {
                if !is_part2 {
                    return score;
                }
                all_seats.extend(seats);
                continue;
            }

            for next_coord @ (next_row, next_col) in neighbors_4(row, col) {
                let next_dr = next_row as isize - row as isize;
                let next_dc = next_col as isize - col as isize;

                if grid[next_row][next_col] == b'#' || next_dr == -dr && next_dc == -dc {
                    continue;
                }

                let next_score = score +
                    if dr == next_dr && dc == next_dc {
                        1
                    } else {
                        1001
                    };

                let key = (next_row, next_col, next_dr, next_dc);

                if *scores.get(&key).unwrap_or(&next_score) >= next_score {
                    scores.insert(key, next_score);

                    let mut seats = seats.clone();
                    seats.push(next_coord);

                    heap.push(Reverse((
                        next_score, next_row, next_col, next_dr, next_dc, seats,
                    )));
                }
            }
        }
        all_seats.len()
    }
}

impl Solution for Day16 {
    const NAME: &'static str = "Reindeer Maze";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        Self::traverse_maze(inp, false)
    }

    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        Self::traverse_maze(inp, true)
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 105496);
        assert_eq!(p2, 524);
    }
}

fn main() {
    aoc_2024::run_day(16, &Day16);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}